File: matlab/src/opt/+opt/EvaluationResult.m
classdef EvaluationResult < handle
    % EvaluationResult  Metrics from evaluating a LINE model via SolverAuto.
    % Mirrors native-Python line_solver.opt.results.EvaluationResult. Per-
    % (station, class) metrics use a containers.Map keyed by 'station||class';
    % utilizations by station; system metrics by chain/class name.

    properties
        feasible = true;
        responseTimes             % Map 'station||class' -> value
        throughputs
        queueLengths
        utilizations              % Map 'station' -> value
        systemResponseTimes       % Map chain/class -> value
        systemThroughputs
        solveTime = 0.0;
        solverUsed = '';
        sensitivities = [];       % opt.SensitivityData or []
    end

    methods
        function obj = EvaluationResult()
            obj.responseTimes = containers.Map('KeyType', 'char', 'ValueType', 'double');
            obj.throughputs = containers.Map('KeyType', 'char', 'ValueType', 'double');
            obj.queueLengths = containers.Map('KeyType', 'char', 'ValueType', 'double');
            obj.utilizations = containers.Map('KeyType', 'char', 'ValueType', 'double');
            obj.systemResponseTimes = containers.Map('KeyType', 'char', 'ValueType', 'double');
            obj.systemThroughputs = containers.Map('KeyType', 'char', 'ValueType', 'double');
        end

        function k = key(~, station, jobclass)
            k = [station '||' jobclass];
        end

        function setResponseTime(obj, station, jobclass, v)
            obj.responseTimes(obj.key(station, jobclass)) = v;
        end
        function setThroughput(obj, station, jobclass, v)
            obj.throughputs(obj.key(station, jobclass)) = v;
        end
        function setQueueLength(obj, station, jobclass, v)
            obj.queueLengths(obj.key(station, jobclass)) = v;
        end

        function v = getResponseTime(obj, station, jobclass)
            if nargin >= 3 && ~isempty(jobclass)
                k = obj.key(station, jobclass);
                if isKey(obj.responseTimes, k), v = obj.responseTimes(k); else, v = inf; end
            else
                v = obj.aggregate(obj.responseTimes, station, true, inf);
            end
        end
        function v = getThroughput(obj, station, jobclass)
            if nargin >= 3 && ~isempty(jobclass)
                k = obj.key(station, jobclass);
                if isKey(obj.throughputs, k), v = obj.throughputs(k); else, v = 0.0; end
            else
                v = obj.aggregate(obj.throughputs, station, false, 0.0);
            end
        end
        function v = getQueueLength(obj, station, jobclass)
            if nargin >= 3 && ~isempty(jobclass)
                k = obj.key(station, jobclass);
                if isKey(obj.queueLengths, k), v = obj.queueLengths(k); else, v = 0.0; end
            else
                v = obj.aggregate(obj.queueLengths, station, false, 0.0);
            end
        end
        function v = getUtilization(obj, station)
            if isKey(obj.utilizations, station), v = obj.utilizations(station); else, v = 0.0; end
        end

        function v = aggregate(~, m, station, doMean, defaultVal)
            ks = keys(m);
            total = 0.0; count = 0;
            prefix = [station '||'];
            for i = 1:numel(ks)
                if strncmp(ks{i}, prefix, numel(prefix))
                    total = total + m(ks{i});
                    count = count + 1;
                end
            end
            if count == 0
                v = defaultVal;
            elseif doMean
                v = total / count;
            else
                v = total;
            end
        end

        function v = getSystemResponseTime(obj, jobclass)
            if nargin >= 2 && ~isempty(jobclass)
                if isKey(obj.systemResponseTimes, jobclass), v = obj.systemResponseTimes(jobclass); else, v = inf; end
                return;
            end
            ks = keys(obj.systemResponseTimes);
            if isempty(ks), v = inf; return; end
            totalTput = 0.0; weighted = 0.0;
            for i = 1:numel(ks)
                rt = obj.systemResponseTimes(ks{i});
                if isKey(obj.systemThroughputs, ks{i}), tp = obj.systemThroughputs(ks{i}); else, tp = 0.0; end
                weighted = weighted + rt * tp; totalTput = totalTput + tp;
            end
            if totalTput > 0
                v = weighted / totalTput;
            else
                vals = cell2mat(values(obj.systemResponseTimes));
                v = sum(vals) / numel(vals);
            end
        end

        function v = getSystemThroughput(obj, jobclass)
            if nargin >= 2 && ~isempty(jobclass)
                if isKey(obj.systemThroughputs, jobclass), v = obj.systemThroughputs(jobclass); else, v = 0.0; end
                return;
            end
            if obj.systemThroughputs.Count == 0, v = 0.0; else, v = sum(cell2mat(values(obj.systemThroughputs))); end
        end
    end
end

File: jar/src/main/java/jline/opt/results/EvaluationResult.java
package jline.opt.results;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Result from evaluating a LINE network model via SolverAUTO, holding the
 * per-(station, class) performance metrics the optimizer reads. Mirrors the
 * native-Python {@code line_solver.opt.results.EvaluationResult}: per-station,
 * per-class metrics are stored as nested station -&gt; class -&gt; value maps
 * (the Java analogue of Python's tuple-keyed dicts), utilizations per station,
 * and system (end-to-end) metrics per chain/class.
 */
public class EvaluationResult {

    public boolean feasible = true;
    // metric kind -> station -> class -> value
    public final Map<String, Map<String, Double>> responseTimes = new LinkedHashMap<String, Map<String, Double>>();
    public final Map<String, Map<String, Double>> throughputs = new LinkedHashMap<String, Map<String, Double>>();
    public final Map<String, Map<String, Double>> queueLengths = new LinkedHashMap<String, Map<String, Double>>();
    public final Map<String, Double> utilizations = new LinkedHashMap<String, Double>();
    public final Map<String, Double> systemResponseTimes = new LinkedHashMap<String, Double>();
    public final Map<String, Double> systemThroughputs = new LinkedHashMap<String, Double>();
    public double solveTime = 0.0;
    public String solverUsed = "";
    /** Analytic sensitivities when product-form; null otherwise. */
    public SensitivityData sensitivities = null;

    private static void put(Map<String, Map<String, Double>> m, String station,
                            String jobclass, double value) {
        Map<String, Double> row = m.get(station);
        if (row == null) {
            row = new LinkedHashMap<String, Double>();
            m.put(station, row);
        }
        row.put(jobclass, value);
    }

    public void setResponseTime(String station, String jobclass, double v) {
        put(responseTimes, station, jobclass, v);
    }

    public void setThroughput(String station, String jobclass, double v) {
        put(throughputs, station, jobclass, v);
    }

    public void setQueueLength(String station, String jobclass, double v) {
        put(queueLengths, station, jobclass, v);
    }

    public double getResponseTime(String station, String jobclass) {
        Map<String, Double> row = responseTimes.get(station);
        if (row == null || !row.containsKey(jobclass)) {
            return Double.POSITIVE_INFINITY;
        }
        return row.get(jobclass);
    }

    /** Aggregate (mean) response time across classes at a station. */
    public double getResponseTime(String station) {
        Map<String, Double> row = responseTimes.get(station);
        if (row == null || row.isEmpty()) {
            return Double.POSITIVE_INFINITY;
        }
        double total = 0.0;
        for (double v : row.values()) {
            total += v;
        }
        return total / row.size();
    }

    public double getThroughput(String station, String jobclass) {
        Map<String, Double> row = throughputs.get(station);
        if (row == null || !row.containsKey(jobclass)) {
            return 0.0;
        }
        return row.get(jobclass);
    }

    public double getThroughput(String station) {
        Map<String, Double> row = throughputs.get(station);
        if (row == null) {
            return 0.0;
        }
        double total = 0.0;
        for (double v : row.values()) {
            total += v;
        }
        return total;
    }

    public double getUtilization(String station) {
        Double v = utilizations.get(station);
        return v == null ? 0.0 : v;
    }

    public double getQueueLength(String station, String jobclass) {
        Map<String, Double> row = queueLengths.get(station);
        if (row == null || !row.containsKey(jobclass)) {
            return 0.0;
        }
        return row.get(jobclass);
    }

    public double getQueueLength(String station) {
        Map<String, Double> row = queueLengths.get(station);
        if (row == null) {
            return 0.0;
        }
        double total = 0.0;
        for (double v : row.values()) {
            total += v;
        }
        return total;
    }

    public double getSystemResponseTime(String jobclass) {
        Double v = systemResponseTimes.get(jobclass);
        return v == null ? Double.POSITIVE_INFINITY : v;
    }

    /**
     * Throughput-weighted mean end-to-end response time across chains, matching
     * total-jobs / total-throughput by Little's law.
     */
    public double getSystemResponseTime() {
        if (systemResponseTimes.isEmpty()) {
            return Double.POSITIVE_INFINITY;
        }
        double totalTput = 0.0;
        double weighted = 0.0;
        for (Map.Entry<String, Double> e : systemResponseTimes.entrySet()) {
            double tput = systemThroughputs.containsKey(e.getKey())
                    ? systemThroughputs.get(e.getKey()) : 0.0;
            weighted += e.getValue() * tput;
            totalTput += tput;
        }
        if (totalTput > 0) {
            return weighted / totalTput;
        }
        double sum = 0.0;
        for (double v : systemResponseTimes.values()) {
            sum += v;
        }
        return sum / systemResponseTimes.size();
    }

    public double getSystemThroughput(String jobclass) {
        Double v = systemThroughputs.get(jobclass);
        return v == null ? 0.0 : v;
    }

    public double getSystemThroughput() {
        double sum = 0.0;
        for (double v : systemThroughputs.values()) {
            sum += v;
        }
        return sum;
    }
}
